contributors/design-proposals/storage/raw-block-pv.md

```diff
@@ -729,7 +728,85 @@ The usage of pod device map path is;
 
 
 
-* unspecified defaults to 'file/ext4' today for backwards compatibility and in mount_linux.go  
+* unspecified defaults to 'file/ext4' today for backwards compatibility and in mount_linux.go
+
+# Dynamically provisioning
+
+Using dynamic provisioning, user is able to create block volume via provisioners. Currently,
+we have two types of provisioners, internal provisioner and external provisioner.
+During volume creation via dynamic provisioner, user passes persistent volume claim which
+contains `volumeMode` parameter, then the persistent volume claim object is passed to
+provisioners. Therefore, in order to create block volume, provisioners need to support
+`volumeMode` and then create persistent volume with `volumeMode`.
+
+If a storage and plugin don't have an ability to create raw block type of volume,
+then `both internal and external provisioner don't need any update` to support `volumeMode`
+because `volumeMode` in PV and PVC are automatically set to `Filesystem` as a default when
+these volume object are created.
```

Comment:
So IIUC, a volume will be created, but because the volumeMode is filesystem, it won't bind to the PVC.  So PVC will stay pending forever.  And then it's up to the user to manually delete the provisioned PVs?  Is there anyway we can propagate some type of error to make it more clear to the user what's wrong?

Reply:
@msau42 @jsafrane 

Good catch. This is a problem, we should provide solution as you mentioned.

Can pv_controller check created volume object after Provision()?
Seems if plugin uses internal provisioner, pv_controller have a chance to inspect whether requested volumeMode in PVC and volumeMode in PV are same or not, so that we can add inspection and show error if they don't match.

But if plugin uses external-provisioner, do we have a chance to inspect PV and show error same as internal provisioner?

Reply:
so not sure I know of an easy way to catch and handle this other than somehow letting the user know via an event that they supplied volumeMode in their pvc but there are currently no supported plugins or matching volumes.  

So in the pv-controller --> index.go, we try to find matchingVolumes based on claimSpecs, etc... We perform this check on misMatchingVolumeModes and maybe we can `flag` that there is a volumeMode mismatch, if that `flag is flipped` on and there are no matches meaning we return:

``` nil, nil (no volume, no error)```

maybe we to return:
```nil, <some error type message>``` 
so the caller can see no volume was returned but an error was returned 

(or another return value indicating that we did find volumeMode being used but did not find a matching volume):
```nil,nil,<some message saying plugin may not support volumeMode or no volumes exist">```

 and maybe then let's put that into an event into the pv and pvc?  

Just thinking outloud?  @msau42 @jsafrane @mtanino  any thoughts?

Reply:
Yeah for external provisioners, probably the best you can do is generate an event if the PV is prebound to the PVC and if volume mode doesn't match.

Reply:
@screeley44 @msau42 

IIUC, the conclusions are;
- For internal provisioner
Check of mismatching volumeMode between PV spec and PVC spec after provision
`pv_controller.go`

- For external provisioner
Check of mismatching volumeMode between PV spec and PVC spec when pv_controller finds matching volumes based on claimSpecs
`index.go`

Reply:
update.

Reply:
+ events on the PVC and the PVC is Pending forever (or until admin fixes either internal or external provisioner)

Reply:
@jsafrane 
Which line are you suggesting to add above comment?

Reply:
It's just for the comment above, the proposal itself is OK.